Don't discard texture bindings for DrawIsland

FalseIlyu opened the pull request:
Should help with #218.
`before`             |  `optimized`
:-------------------------:|:-------------------------:
![before](https://github.com/openblack/openblack/assets/83550517/568dd12c-8b67-4413-ad3d-1823b9b3c0b7)|![optimized](https://github.com/openblack/openblack/assets/83550517/0f91c4ed-3156-4c08-9605-e9090e4e6dc2)

bwrsandman commented:
Can you explain more in detail what you found and what you did

FalseIlyu commented:
> Can you explain more in detail what you found and what you did

When Drawing the island we rebind the same texture for each Island Block. By setting bgfx to not discard texture bindings on submit we can bind the textures at the start and keep the same bindings for all blocks.

bwrsandman commented:
Do you plan on implementing #81 ? I don't think this change is necessary if we do instanced drawing

FalseIlyu commented:
> Looks well done. However, do you plan on implementing #81 ? I don't think this change is necessary if we do instanced drawing

I don't know, that's why i took the low hanging fruit in term of performance gain.

bwrsandman commented:
I think what this PR demonstrate that the states are maintained throughout the island draws. That would mean that #81 will be fairly trivial to implement.

bwrsandman reviewed (changes requested):
Great work. However in order for this to be complete, I think converting the island draws to instanced drawing would be the correct solution.

FalseIlyu commented:
> I think what this PR demonstrate that the states are maintained throughout the island draws. That would mean that #81 will be fairly trivial to implement.

I don't know making the shader handle altitude is trivial, but moving material blending to shader/blenderer pretty much needs to rewrite the way terrain is drawn to make instance of cells and not block (and i see it as fairly not trivial) or get the data for all cell in a block to the shader (which doesn't seem that trivial either).

bwrsandman commented:
Did you mean to add two commits about instance buffers?

FalseIlyu commented:
> Did you mean to add two commits about instance buffers?

No i did not, sorry my mistake.

bwrsandman commented:
This commit (6f7844708ab0628cc5ab4ac5eb5fe23281937608) has some instancing code but as far as I can tell is not drawing instances.
![image](https://github.com/openblack/openblack/assets/1013356/82226c76-202b-47ee-bb77-264c3ba603ae)

FalseIlyu commented:
> This commit ([6f78447](https://github.com/openblack/openblack/commit/6f7844708ab0628cc5ab4ac5eb5fe23281937608)) 
has some instancing code but as far as I can tell is not drawing instances.

We draw block by block but we send u_mapPositionAndSize through an instance buffer that we reallocate at each block (it's still faster than updating the uniform); i guess a better way to do it would be to move block size to u_islandConstant and blockPosition to the Vertex data.

bwrsandman commented:
> instance buffer that we reallocate at each block

That's not how instance buffers are used and re-allocating multiple times a frame is very bad for pipeline since you're interrupting the GPU and you may cause race conditions.

FalseIlyu commented:
> > instance buffer that we reallocate at each block
> 
> That's not how instance buffers are used. They're meant to be indexed using the the instance id. Also, re-allocating multiple times a frame is very bad for pipelining since you're interrupting the GPU and you may cause race conditions. The proper way to use them is to fill them up for each instance as an array of data before you start submitting draw calls.

I see the problem, i don't see how to implement the solution without basically doing #615 so i just rebased (to make this compatible with #616) and dropped this change. I think we still improve the performance significantly this way.

bwrsandman commented:
Why are you sampling the height map in this changeset?

FalseIlyu commented:
> Why are you sampling the height map in this changeset? Wouldn't that be a #615 thing?

It may have been done in the time before #615 became 

> Why are you sampling the height map in this changeset? Wouldn't that be a #615 thing?

It is a #615 thing, i'm reverting it.

bwrsandman commented:
Can the rest of the changes in `vs_terrain.sc` be removed to have a minmal changeset?

FalseIlyu commented on src/Renderer.cpp:
```diff
@@ -548,38 +548,49 @@ void Renderer::DrawPass(const DrawSceneDesc& desc) const
-					| BGFX_STATE_MSAA
-				;
-				// clang-format on
+				terrainShader->SetUniformValue("u_mapPositionAndSize", &mapPositionAndSize);
```
Thanks, i'm on it
